pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go

```diff
@@ -242,52 +239,63 @@ func (n *klusterletController) sync(ctx context.Context, controllerContext facto
 		appliedManifestWorkClient: n.appliedManifestWorkClient,
 	}
 
-	// If there are some invalid feature gates of registration, will output condition `ValidRegistrationFeatureGates` False in Klusterlet.
-	if klusterlet.Spec.RegistrationConfiguration != nil && len(klusterlet.Spec.RegistrationConfiguration.FeatureGates) > 0 {
-		featureGateArgs, invalidFeatureGates := helpers.FeatureGatesArgs(
-			klusterlet.Spec.RegistrationConfiguration.FeatureGates, helpers.ComponentSpokeRegistrationKey)
-		if len(invalidFeatureGates) == 0 {
-			config.RegistrationFeatureGates = featureGateArgs
-			// TODO: after the 0.10.0 is released, change back to UpdateKlusterletConditionFn
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.ReplaceKlusterletConditionFn(
-				spokeRegistrationFeatureGatesInvalid,
-				metav1.Condition{
-					Type: spokeRegistrationFeatureGatesValid, Status: metav1.ConditionTrue, Reason: "FeatureGatesAllValid",
-					Message: "Registration feature gates of klusterlet are all valid",
-				}))
-		} else {
-			invalidFGErr := fmt.Errorf("there are some invalid feature gates of registration: %v ", invalidFeatureGates)
-			// TODO: after the 0.10.0 is released, change back to UpdateKlusterletConditionFn
-			_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.ReplaceKlusterletConditionFn(
-				spokeRegistrationFeatureGatesInvalid,
-				metav1.Condition{
-					Type: spokeRegistrationFeatureGatesValid, Status: metav1.ConditionFalse, Reason: "InvalidFeatureGatesExisting",
-					Message: invalidFGErr.Error(),
-				}))
-			return invalidFGErr
+	// Do not use "conditions := &klusterlet.Status.Conditions", otherwise will override conditions
+	// set by other controllers
+	conditions := &[]metav1.Condition{}
+	var statusUpdated bool = false
+	defer func() {
+		klog.V(2).Infof("Update Klusterlet %q status, statusUpdated: %v", klusterletName, statusUpdated)
+		if statusUpdated {
+			return
+		}
+
+		_, _, updateErr := helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName,
+			helpers.UpdateKlusterletConditionFn(*conditions...))
+		if updateErr != nil {
+			klog.V(2).Infof("Update Klusterlet %q status with error: %v", klusterletName, updateErr)
+		}
+	}()
+
+	// If there are some invalid feature gates of registration, will output condition
+	// `ValidRegistrationFeatureGates` False in Klusterlet.
+	if klusterlet.Spec.RegistrationConfiguration != nil &&
+		len(klusterlet.Spec.RegistrationConfiguration.FeatureGates) > 0 {
```

Comment:
we should consider empty case

Reply:
so, if the feature gates are empty, do you expect to set a valid condition?

Reply:
yes